index.html

```diff
@@ -126,187 +846,604 @@ <h3>
+            set to |pullAlgorithm|,
+            <a href="https://streams.spec.whatwg.org/#readablestream-set-up-cancelalgorithm">cancelAlgorithm</a>
+            set to |cancelAlgorithm|,
+            <a href="https://streams.spec.whatwg.org/#readablestream-set-up-highwatermark">highWaterMark</a>
+            set to [=this=].{{UDPSocketOptions/readableStreamBufferSize}}.
+          <li>Set [=this=].{{UDPSocket/[[readable]]}} to |stream|.
+        </ol>
+
+        To <dfn>handle closing the UDP readable stream</dfn> perform the following steps:
+
+        <ol>
+          <li>
+            If [=this=].{{UDPSocket/[[writable]]}} is errored:
+            <ul>
+              <li> 
+                [=Reject=] [=this=].{{UDPSocket/[[closedPromise]]}}.
+            </ul>
+          <li> 
+            If [=this=].{{UDPSocket/[[writable]]}} is closed:
+            <ul>
+              <li>
+                If [=this=].{{UDPSocket/[[readable]]}} is closed, [=resolve=] [=this=].{{UDPSocket/[[closedPromise]]}} with `undefined`.
+              <li>
+                If [=this=].{{UDPSocket/[[readable]]}} is errored, [=reject=] [=this=].{{UDPSocket/[[closedPromise]]}}.
+            </ul>
+        </ol>
+
+        <section data-dfn-for="UDPMessage">
+
+          <h4><dfn>UDPMessage</dfn> dictionary</h4>
+
           <pre class="idl">
-          dictionary TCPSocketOptions : SocketOptions {
-            boolean noDelay;
-            boolean keepAlive;
-            [EnforceRange] unsigned long keepAliveDelay;
-          };
+            dictionary UDPMessage {
+              BufferSource   data;
+              DOMString      remoteAddress;
+              unsigned short remotePort;
+            };
           </pre>
-          <p>
-            The <dfn>{{TCPSocketOptions}}</dfn> dictionary inherits all attributes 
-            from {{SocketOptions}} and exposes optional TCP-specific options.
-          </p>
+
           <dl>
             <dt>
-              <dfn>noDelay</dfn> member
+              <dfn>data</dfn> member
             </dt>
             <dd>
-              Enables the `TCP_NODELAY` option,
-              disabling <a href=
-              "https://en.wikipedia.org/wiki/Nagle%27s_algorithm">Nagle's
-              algorithm</a>.
+              The user message represented as {{BufferSource}}.
+              <div class="note">
+                Note that for {{UDPSocket/[[readable]]}} the underlying type is always {{Uint8Array}}.
+              </div>
             </dd>
             <dt>
-              <dfn>keepAlive</dfn> member
+              <dfn>remoteAddress</dfn> member
             </dt>
             <dd>
-              Enables/disables TCP Keep-Alive by setting `SO_KEEPALIVE` option on the socket. 
+              The remote address where the message came from / where the message should be sent.
+              <div class="note">
+                The only current implementation (Chromium) ignores this parameter when writing.
```

Comment:
Backwards compatibility is quite an interesting question here -- I would assume that the easiest way to ensure this is to replace `remoteAddress` / `remotePort` with `address` / `port` and introduce a `mode` field to the options some time in the future. This way we'll allow developers to operate on bound sockets as well without creating unconnected sockets.

Reply:
Replacing `remoteAddress` / `remotePort` with `address` / `port` makes sense to me because sendto(2) and recvfrom(2) only accept or provide the remote address. Setting the local address requires calling bind(2).

It looks like even after calling connect(2) it is still possible to give sendto(2) a different address. Is there any reason not to specify and implement these options to sendto(2) in the initial version?

Reply:
`Send(...)` requires `Connect(...)` but `SendTo(...)` only requires `Bind(...)`. As specified you always call `Connect(...)` with the address and port specified in the constructor but you could make those constructor parameters optional.

Reply:
Let's resolve this for being a duplicate for [discussion_r921586730](https://github.com/WICG/direct-sockets/pull/48#discussion_r921586730).